cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java

```diff
@@ -866,10 +867,18 @@ public void searchMetadata(HttpRequest request, HttpResponder responder,
         throw new BadRequestException("Cursors are not supported when sort info is not specified.");
       }
     }
-    MetadataSearchResponse response =
-      metadataAdmin.search(namespaceId, URLDecoder.decode(searchQuery, "UTF-8"), types,
-                           sortInfo, offset, limit, numCursors, cursor, showHidden);
-    responder.sendJson(HttpResponseStatus.OK, response, MetadataSearchResponse.class, GSON);
+    try {
+      MetadataSearchResponse response =
+        metadataAdmin.search(namespaceId, URLDecoder.decode(searchQuery, "UTF-8"), types,
+                             sortInfo, offset, limit, numCursors, cursor, showHidden);
+      responder.sendJson(HttpResponseStatus.OK, response, MetadataSearchResponse.class, GSON);
+    } catch (Exception e) {
+      // if MetadataDataset throws an exception, it gets wrapped
+      if (Throwables.getRootCause(e) instanceof IllegalArgumentException) {
+        throw new BadRequestException(e);
```

Comment:
I agree with the comment on the original PR: the BadRequestException should have the message of the IllegalArgumentException. Otherwise the HTTP response will have an empty (or null?) message.